pkgs/development/compilers/tinygo/default.nix

```
@@ -1,64 +1,129 @@
+    (substituteAll {
+      src = ./0001-Add-clang-header-path.patch;
+      clang_include = "${clang.cc.lib}/lib/clang/${clang.cc.version}/include";
+    })
+    ./0002-Makefile.patch
+  ];
 
-  preBuild = "cp ${gomod} go.mod";
+  vendorSha256 = "sha256-fK8BlCh+1NtHW6MwW68iSIB+Sw6AK+g3y4lMyMYrXkk=";
 
-  postBuild = "make gen-device";
+  nativeBuildInputs = [ makeWrapper ];
+  buildInputs = [ llvm libffi clang.cc ];
+  checkInputs = [ avrgcc binaryen ];
 
-  vendorSha256 = "12k2gin0v7aqz5543m12yhifc0xsz26qyqra5l4c68xizvzcvkxb";
+  doCheck = stdenv.buildPlatform == stdenv.hostPlatform;
 
-  doCheck = false;
+  subPackages = [ "." ];
+  tags = [ "llvm${llvmMajor}" ];
+  allowGoReference = true;
 
-  prePatch = ''
-    sed -i s/', "-nostdlibinc"'// builder/builtins.go
-    sed -i s/'"-nostdlibinc", '// compileopts/config.go builder/picolibc.go
-  '';
+  # only strip the compiler
+  stripDebugList = [ "bin" ];
 
-  subPackages = [ "." ];
-  nativeBuildInputs = [ makeWrapper ];
-  buildInputs = [ llvm clang-unwrapped ];
-  propagatedBuildInputs = [ lld avrgcc avrdude openocd gcc-arm-embedded ];
-
-  postInstall = ''
-    mkdir -p $out/share/tinygo
-    cp -a lib src targets $out/share/tinygo
-    wrapProgram $out/bin/tinygo --prefix "TINYGOROOT" : "$out/share/tinygo" \
-      --prefix "PATH" : "$out/libexec/tinygo"
+  postPatch = ''
+    # Copy wasi-libc, symlink seems not working
+    rm -rf lib/wasi-libc/*
+    mkdir -p lib/wasi-libc/sysroot/lib/wasm32-wasi lib/wasi-libc/sysroot/include
+    cp -a ${wasi-libc}/lib/* lib/wasi-libc/sysroot/lib/wasm32-wasi/
+    cp -a ${wasi-libc.dev}/include/* lib/wasi-libc/sysroot/include/
+
+    # Borrow compiler-rt builtins from our source
+    # See https://github.com/tinygo-org/tinygo/pull/2471
+    mkdir -p lib/compiler-rt-builtins
+    cp -a ${compiler-rt.src}/compiler-rt/lib/builtins/* lib/compiler-rt-builtins/
+
+    substituteInPlace Makefile \
+      --replace "\$(TINYGO)" "/build/source/build/tinygo" \
+      --replace "build/release/tinygo/bin" "$out/bin" \
+      --replace "build/release/" "$out/share/"
+
+    #TODO: Fix mingw and darwin
+    # Disable windows and darwin cross-compile tests
+    sed -i "/GOOS=windows/d" Makefile
+    sed -i "/GOOS=darwin/d" Makefile
+
+    # tinygo needs versioned binaries
     mkdir -p $out/libexec/tinygo
-    ln -s ${clang-unwrapped}/bin/clang $out/libexec/tinygo/clang-10
-    ln -s ${lld}/bin/lld $out/libexec/tinygo/ld.lld-10
-    ln -sf $out/bin $out/share/tinygo
+    ln -s ${lib.getBin clang.cc}/bin/clang $out/libexec/tinygo/clang-${llvmMajor}
+    ln -s ${lib.getBin lld}/bin/ld.lld $out/libexec/tinygo/ld.lld-${llvmMajor}
+    ln -s ${lib.getBin lld}/bin/wasm-ld $out/libexec/tinygo/wasm-ld-${llvmMajor}
+    ln -s ${gdb}/bin/gdb $out/libexec/tinygo/gdb-multiarch
+  '';
+
+  postBuild = ''
+    # Move binary
+    mkdir build
+    mv $GOPATH/bin/tinygo build/tinygo
+
+    make gen-device
+  '';
+
+  # Avoid "error: mkdir /homeless-shelter: permission denied"
```

```suggestion
  # Tries to write to $HOME
```

Also this should be exporter in some pre phase.

---

Our `buildGoModule` exports `GOCACHE` but `tinygo` doesn't respect that :/